lib/spear/connection.ex

```diff
@@ -115,22 +116,94 @@ defmodule Spear.Connection do
   defp process_response(_unknown, state), do: state
 
   defp stream_body(state, request_ref, messages) do
-    Enum.reduce_while(messages, {:ok, state}, &stream_body_message(&1, &2, request_ref))
+    Enum.reduce_while(
+      messages,
+      {:ok, state},
+      &stream_body_message(&1, &2, request_ref)
+    )
   end
 
   defp stream_body_message(message, {:ok, state}, request_ref) do
-    {wire_data, _byte_size} = Request.to_wire_data(message)
-
-    stream_result =
-      Mint.HTTP.stream_request_body(
-        state.conn,
-        request_ref,
-        wire_data
-      )
-
-    case stream_result do
-      {:ok, conn} -> {:cont, {:ok, put_in(state.conn, conn)}}
-      {:error, conn, reason} -> {:halt, {:error, put_in(state.conn, conn), reason}}
+    {wire_data, byte_size} = Request.to_wire_data(message)
+    smallest_window = get_smallest_window(state.conn, request_ref)
+
+    with false <- byte_size > smallest_window,
+         {:ok, conn} <- Mint.HTTP.stream_request_body(state.conn, request_ref, wire_data) do
+      {:cont, {:ok, put_in(state.conn, conn)}}
+    else
+      _window_too_small? = true ->
+        recv_until_window_size_increase(state, smallest_window, request_ref)
+
+      {:error, conn, reason} ->
+        {:halt, {:error, put_in(state.conn, conn), reason}}
+    end
+  end
+
+  defp recv_until_window_size_increase(state, current_window, request_ref) do
+    params = {current_window, request_ref}
+
+    {:ok, state}
+    |> do_stage(:recv_responses, params)
+    |> do_stage(:check_window_size, params)
```

Comment:
this is the flow-control mechanism I was describing

initially this block of code for `recv_until_window_size_increase/3` was more complicated because I was trying to switch the `conn` into passive mode and then clean up by switching it back into active mode, but `Mint.HTTP.set_mode/2` has a return signature of `set_mode(conn, :active | :passive) :: {:ok, conn()} | {:error, any()}` (or something like that), so even that
had some tuple-handling to do

it's probably possible to refactor this function now into a `with` or two